prow/cmd/deck/job_history.go

```diff
@@ -0,0 +1,389 @@
+		}
+		if buildID < 0 {
+			return bucketName, root, buildID, fmt.Errorf("invalid value %s = %d", idParam, buildID)
+		}
+	}
+
+	return bucketName, root, buildID, nil
+}
+
+func linkID(url *url.URL, id int64) string {
+	u := *url
+	q := u.Query()
+	var val string
+	if id != emptyID {
+		val = strconv.FormatInt(id, 10)
+	}
+	q.Set(idParam, val)
+	u.RawQuery = q.Encode()
+	return u.String()
+}
+
+func getBuildData(bkt *storage.BucketHandle, root string, buildID int64, index int) buildData {
+	b := buildData{
+		index:  index,
+		ID:     strconv.FormatInt(buildID, 10),
+		Result: "Unfinished",
+	}
+	link, err := spyglassLink(bkt, root, b.ID)
+	if err != nil {
+		logrus.Warning(err)
+		return b
+	}
+	b.SpyglassLink = link
+	started := jobs.Started{}
+	err = readJSON(bkt, root, b.ID, "started.json", &started)
+	if err == nil {
+		b.Started = time.Unix(started.Timestamp, 0)
+		finished := jobs.Finished{}
+		readJSON(bkt, root, b.ID, "finished.json", &finished)
+		if finished.Timestamp != 0 {
+			b.Duration = time.Unix(finished.Timestamp, 0).Sub(b.Started)
+		}
+		if finished.Result != "" {
+			b.Result = finished.Result
+		}
+	} else {
+		logrus.Warning(err)
+	}
+	return b
+}
+
+// assumes a to be sorted in descending order
+// returns a subslice of a along with its indices (inclusive)
+func cropResults(a []int64, max int64) ([]int64, int, int) {
+	res := []int64{}
+	firstIndex := -1
+	lastIndex := 0
+	for i, v := range a {
+		if v <= max {
+			res = append(res, v)
+			if firstIndex == -1 {
+				firstIndex = i
+			}
+			lastIndex = i
+			if len(res) >= resultsPerPage {
+				break
+			}
+		}
+	}
+	return res, firstIndex, lastIndex
+}
+
+// golang <3
+type int64slice []int64
+
+func (a int64slice) Len() int           { return len(a) }
+func (a int64slice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a int64slice) Less(i, j int) bool { return a[i] < a[j] }
+
+// Gets job history from the GCS bucket specified in config.
+func getJobHistory(url *url.URL, config *config.Config, gcsClient *storage.Client) (jobHistoryTemplate, error) {
+	start := time.Now()
+	tmpl := jobHistoryTemplate{}
+
+	bucketName, root, top, err := jobHistURL(url)
+	if err != nil {
+		return tmpl, fmt.Errorf("invalid url %s: %v", url.String(), err)
+	}
+	tmpl.Name = root
+	bkt := gcsClient.Bucket(bucketName)
+
+	latest, err := readLatestBuild(bkt, root)
+	if err != nil {
+		return tmpl, fmt.Errorf("failed to locate build data: %v", err)
+	}
+	if top == emptyID || top > latest {
+		top = latest
+	}
+	if top != latest {
+		tmpl.LatestLink = linkID(url, emptyID)
+	}
+
+	buildIDs, err := listBuildIDs(bkt, root)
+	if err != nil {
+		return tmpl, fmt.Errorf("failed to get build ids: %v", err)
+	}
+	sort.Sort(sort.Reverse(int64slice(buildIDs)))
+
+	shownIDs, firstIndex, lastIndex := cropResults(buildIDs, top)
+	if firstIndex > 0 {
+		prevIndex := firstIndex - resultsPerPage
+		prev := emptyID
+		if prevIndex > 0 {
```

Comment:
When a job history URL lacks a build ID, the newest build is displayed at the top. Here I use `emptyID` instead of `buildIDs[0]` because the newest build might not exist yet.